api/v1alpha1/reclaimspacejob_types.go

```diff
@@ -0,0 +1,114 @@
+/*
+Copyright 2021 The Kubernetes-CSI-Addons Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// OperationState defines the state of the operation.
+type Result string
+
+const (
+	// ResultSucceeded represents the Succeeded operation state.
+	ResultSucceeded Result = "Succeeded"
+
+	// ResultFailed represents the Failed operation state.
+	ResultFailed Result = "Failed"
+)
+
+// TargetSpec defines the targets on which the operation can be
+// performed.
+type TargetSpec struct {
+	// PVC represents PersistentVolumeClaim name.
+	PVC string `json:"pvc,omitempty"`
+}
+
+// ReclaimSpaceJobSpec defines the desired state of ReclaimSpaceJob
+type ReclaimSpaceJobSpec struct {
+	// Target represents volume target on which the operation will be
+	// performed.
+	// +kubebuilder:validation:Required
+	Target TargetSpec `json:"target"`
+
+	// BackOffLimit specifies the number of retries allowed before marking reclaim
+	// space operation as failed. If not specified, defaults to 6.
+	// +optional
+	// +kubebuilder:default:=6
+	BackoffLimit int32 `json:"backOffLimit"`
+
+	// ActiveDeadlineSeconds specifies the duration in seconds relative to the
+	// start time that the operation might be retried; value must be positive integer.
+	// If not specified, defaults to 600 seconds.
+	// +optional
+	// +kubebuilder:default:=600
+	ActiveDeadlineSeconds int64 `json:"activeDeadlineSeconds"`
```

Comment:
i was discussing this with @Rakshith-R a couple of hours back. do we need this one? as we are not calling any RPC calls does this helps in any way

Reply:
> i was discussing this with @Rakshith-R a couple of hours back. do we need this one? as we are not calling any RPC calls does this helps in any way

Changed it to `RetryDeadlineSeconds`, IMO, it will be helpful to limit the retry of operation based on duration from the start of operation.

Similar to how a job is killed if it exceeds `ActiveDeadlineSeconds`
 https://pkg.go.dev/k8s.io/api/batch/v1#JobSpec

Reply:
but we are not going to kill the running fstrim/ operations here but we just stop the reconcile.

Reply:
>but we are not going to kill the running fstrim/ operations here but we just stop the reconcile.

Yes, We will be stopping the retries here.

Reply:
am not sure this really helps. if it was killing/stopping the operation this makes sense.

Reply:
> am not sure this really helps. if it was killing/stopping the operation this makes sense

It's just killing/stopping the next retry of the operation, both results in making sure the operation does not exceed time duration limit, one is less restrict than another. I believe we do need a way to limit the execution of operation to a particular approximate time period.

Reply:
am still not convinced for handing this deadline seconds @nixpanic what is your thought on this one? i don't have anything else except this one.

Reply:
`ActiveDeadlineSeconds` sounds reasonable to me. I expect the operation to be in a pending/active state until it either returns a response or until this timeout expires. Upon expiration of the timeout, an error can be returned to the caller, and the retry logic can be applied. At this point the operation is not active anymore.

There is no need to _kill_ the running operation in this case, the deadline is most likely hit when the side-car or csi-driver faces an issue (restart, network troubles, ...).

Reply:
```

	// RetryDeadlineSeconds specifies the duration in seconds relative to the
	// start time that the operation may be retried; value MUST be positive integer.
	// If not specified, defaults to 600 seconds. Maximum allowed
	// value is 1800.
	// +optional
	// +kubebuilder:validation:Maximum=1800
	// +kubebuilder:validation:Minimum=0
	// +kubebuilder:default:=600
	RetryDeadlineSeconds int64 `json:"activeDeadlineSeconds"`
```

The latest iteration has `RetryDeadlineSeconds`, which is checked at the start of reconcile only. 

We'll not be using this to set timeouts at sidecar or driver level.

This is wholly to restrict operation to within a limited duration. Backofflimit only limits number of retries.

It is also designed to work in addition to `startDeadlineSeconds` of CronJob, this will make sure the operation will not be performed outside `StartDeadlineSeconds + RetryDeadlineSeconds`.
(Similar to k8s cronjob)

This will ensure reclaimspace io intensive operation does not slip outside maintenance period.

@Madhu-1 @nixpanic

Reply:
> This will ensure reclaimspace io intensive operation does not slip outside maintenance period.

:point_up: is difficult to achieve. Once the operation is started, it is not really possible to cancel/kill it (except when killing the csi-driver pod doing the operation). My understanding is that this parameter is an _assumed to be inactive when no response is received after the deadline_ timer, just so that the retry logic can kick in.

`RetryDeadlineSeconds` or `ActiveDeadlineSeconds` both work for me, as long as I understand what the option should do.

Reply:
We can discuss more during the implementation of the reclaimspacejob & reclaimspacecronjob controllers and make modifications accordingly to the API later too. 
The usage can be better understood from the controllers' implementation. 

This sounds okay to you @Madhu-1 ?